test/karma/test-app/util.ts

```diff
@@ -100,39 +172,62 @@ export function setupDomTests(document: Document) {
   }
 
   /**
-   * Create web component for executing tests against
+   * Render HTML for executing tests against.
+   * @param url the location on disk containing the HTML to load
+   * @param testElement a parent HTML element to place test code in
+   * @param waitForStencilReady the amount of time to wait (in milliseconds) to wait for an application to load
+   * @returns the fully rendered HTML to test against
    */
-  function renderTest(url: string, app: HTMLElement, waitForStencilReady: number) {
+  function renderTest(url: string, testElement: HTMLElement, waitForStencilReady?: number) {
     url = path.join('base', WWW_OUT_DIR, url);
 
     return new Promise<HTMLElement>((resolve, reject) => {
       try {
-        const allReady = () => {
-          const promises: Promise<any>[] = [];
-          const waitForDidLoad = (promises: Promise<any>[], elm: Element) => {
+        /**
+         * Ensure that all `onComponentReady` functions on Stencil elements in the DOM have been called before rendering
+         * @returns an array of promises, one for each `onComponentReady` found on a Stencil component
+         */
+        const allReady = (): Promise<d.HTMLStencilElement[] | void> => {
+          const promises: Promise<d.HTMLStencilElement>[] = [];
+          /**
+           * Function that recursively traverses the DOM, looking for Stencil components. Any `componentOnReady`
+           * functions found on Stencil components are pushed to a buffer to be run after traversing the entire DOM.
+           * @param elm the current element being inspected
+           */
+          const waitForDidLoad = (elm: Element): void => {
             if (elm != null && elm.nodeType === 1) {
+              // the element exists and is an `ELEMENT_NODE`
               for (let i = 0; i < elm.children.length; i++) {
```

Comment:
Yeah, I'm seeing that same behavior too - where `for (let el of document.body.children) { console.log(el) }` works for me.  Specifically, the type system is what's disallowing that for us:
<img width="616" alt="Screen Shot 2022-07-11 at 1 59 10 PM" src="https://user-images.githubusercontent.com/1930213/178327980-a1098296-d9f0-46b1-8a9b-a5db68d18ab1.png">

Reply:
just played around with this a little bit, this is a bit funky and I don't quite get why typescript is like this, but even though DOM collection objects like `HTMLCollection` should be iterable since ES6 you have to add an additional lib file `DOM.Iterable` to `tsconfig.json` in order to get those typings. I was able to get the error shown in the screenshot to stop showing up by doing the following:

```diff
diff --git a/test/karma/tsconfig.json b/test/karma/tsconfig.json
index ba82ac6b8..731e0b733 100644
--- a/test/karma/tsconfig.json
+++ b/test/karma/tsconfig.json
@@ -11,7 +11,8 @@
     "jsxFactory": "h",
     "lib": [
       "dom",
-      "es2017"
+      "es2017",
+      "DOM.Iterable"
     ],
     "module": "commonjs",
     "moduleResolution": "node",
```

I'm not sure if we want to make that change or not, but it does work with that!

Reply:
Hmmm interestingly, I can't seem to make the type system happy with the diff above.  Take a look at the changes I had started working on, LMK if you can get it to build with this (FWIW, I was running `npm run clean && npm run build.all`):

```diff
diff --git a/test/karma/test-app/util.ts b/test/karma/test-app/util.ts
index 59f5ea520..6be9ee62d 100644
--- a/test/karma/test-app/util.ts
+++ b/test/karma/test-app/util.ts
@@ -197,8 +197,7 @@ export function setupDomTests(document: Document): DomTestUtilities {
           const waitForDidLoad = (elm: Element): void => {
             if (elm != null && elm.nodeType === 1) {
               // the element exists and is an `ELEMENT_NODE`
-              for (let i = 0; i < elm.children.length; i++) {
-                const childElm = elm.children[i];
+              for (const childElm of elm.children) {
                 if (childElm.tagName.includes('-') && isHtmlStencilElement(childElm)) {
                   promises.push(childElm.componentOnReady());
                 }
@@ -257,28 +256,28 @@ export function setupDomTests(document: Document): DomTestUtilities {
            */
           const parseAndRebuildScriptTags = () => {
             const tempScripts: NodeListOf<HTMLScriptElement> = testElement.querySelectorAll('script');
-            for (let i = 0; i < tempScripts.length; i++) {
+            for (const tempScript of tempScripts) {
               const script: HTMLScriptElement = document.createElement('script');
-              if (tempScripts[i].src) {
-                script.src = tempScripts[i].src;
+              if (tempScript.src) {
+                script.src = tempScript.src;
               }
-              if (tempScripts[i].hasAttribute('nomodule')) {
+              if (tempScript.hasAttribute('nomodule')) {
                 script.setAttribute('nomodule', '');
               }
-              if (tempScripts[i].hasAttribute('type')) {
-                const typeAttribute = tempScripts[i].getAttribute('type');
+              if (tempScript.hasAttribute('type')) {
+                const typeAttribute = tempScript.getAttribute('type');
                 if (typeof typeAttribute === 'string') {
                   // older DOM implementations would return an empty string to designate `null`
                   // here, we interpret the empty string to be a valid value
                   script.setAttribute('type', typeAttribute);
                 }
               }
-              script.innerHTML = tempScripts[i].innerHTML;
+              script.innerHTML = tempScript.innerHTML;
 
-              if (tempScripts[i].parentNode) {
+              if (tempScript.parentNode) {
                 // the scripts were found by querying a common parent node, which _should_ still exist
-                tempScripts[i].parentNode!.insertBefore(script, tempScripts[i]);
-                tempScripts[i].parentNode!.removeChild(tempScripts[i]);
+                tempScript.parentNode!.insertBefore(script, tempScript);
+                tempScript.parentNode!.removeChild(tempScript);
               } else {
                 // if for some reason the parent node no longer exists, something's manipulated it while we were parsing
                 // the script tags. this can lead to undesirable & hard to debug behavior, fail.
diff --git a/test/karma/tsconfig.json b/test/karma/tsconfig.json
index ba82ac6b8..d2536f597 100644
--- a/test/karma/tsconfig.json
+++ b/test/karma/tsconfig.json
@@ -4,6 +4,7 @@
     "allowSyntheticDefaultImports": true,
     "allowUnreachableCode": true,
     "declaration": false,
+    "downlevelIteration": true,
     "resolveJsonModule": true,
     "experimentalDecorators": true,
     "forceConsistentCasingInFileNames": true,
@@ -11,7 +12,8 @@
     "jsxFactory": "h",
     "lib": [
       "dom",
-      "es2017"
+      "es2017",
+      "DOM.Iterable"
     ],
     "module": "commonjs",
     "moduleResolution": "node",
@@ -20,7 +22,7 @@
     "noUnusedLocals": false,
     "noUnusedParameters": false,
     "pretty": true,
-    "target": "es5",
+    "target": "es6",
     "useUnknownInCatchVariables": true,
     "outDir": "tmp-compiled-tests",
     "baseUrl": ".",
```

Reply:
I bumped the target version above as a dumb check to see if that would help the compilation errors go away 😆 Ideally, I don't want to tweak too much of the tsconfig as a part of this PR - the original intent here was just to get some documentation/better structure around the util file after I realized that wasn't causing our test flakiness

Reply:
hmmmm yeah, maybe we should just not worry about this for now...